Improve UCX documentation and examples

charlesbluca opened the pull request:
Addressing #544, this PR aims to clarify the requirements, configuration, and usage of UCX with Dask-CUDA.

Still a lot to be done:

- [x] Flesh out hardware/software requirements
- [x] Rework CLI/Python usage examples
- [x] Clarify some uncertainties in the Configuration section
- [ ] Add standalone examples of UCX usage

charlesbluca commented:
Thinking that the [UCX usage sections](https://dask-cuda.readthedocs.io/en/latest/ucx.html#launching-scheduler-workers-and-clients-separately) could probably be simplified a bit if we better fleshed out `dask-cuda-worker` and `dask_cuda.initialize(...)` in their own pages.

pentschev commented on docs/source/ucx.rst:
```diff
@@ -1,35 +1,87 @@
+    Explicitly sets ``UCX_NET_DEVICES`` instead of defaulting to ``"all"``, which can result in suboptimal performance.
+    If using InfiniBand, set to the desired IB device, e.g. ``"mlx5_0:1"``.
+    If InfiniBand is disabled, set to the ethernet device, e.g. ``"enp1s0f0"`` on a DGX-1.
+    All available UCX-compatible devices can be listed by running ``ucx_info -d`` or ``ifconfig`` *(are all the options under ifconfig compatible)*.
```
I don't think all options under `ifconfig` are valid, for example `veth*` (often used by Docker) aren't. `ucx_info -d` only should be more reliable.

Also, interestingly the most important use case for us is missing in this section (it was missing before your proposed changes here too), which is the `"auto"` option, as documented in https://github.com/rapidsai/dask-cuda/blob/e72d776645ee3ee7e1672f73c9c043731cf401d5/dask_cuda/cli/dask_cuda_worker.py#L187-L203 and https://github.com/rapidsai/dask-cuda/blob/e72d776645ee3ee7e1672f73c9c043731cf401d5/dask_cuda/local_cuda_cluster.py#L72-L86 . Could you add that here as well? It's documented in "Important Notes" in this file below, but it seems sensible to have it here too.

charlesbluca replied:
It does make sense to include information about `"auto"` here. Are there any resources on installing UCX-Py with hwloc support? Didn't see anything explicit about it in the installation guide.

pentschev replied:
It's in the build dependencies: https://ucx-py.readthedocs.io/en/latest/install.html#build-dependencies

charlesbluca replied:
Question about specifying an interface when InfiniBand is disabled:

I notice later on in the documentation, we say when setting up a scheduler for UCX that setting `DASK_UCX__NET_DEVICES` can generally be skipped when using a non-InfiniBand interface is used. Is this universally true, or does it only apply to the scheduler?

pentschev replied:
This can apply to all processes. `UCX_NET_DEVICES` is merely a way to limit the visibility of network devices to UCX, meaning it won't "see" any others. We can still choose the device by specifying `--interface` to scheduler and workers, which is generally preferred if no IB is used. To be even more explicit, we're likely going to drop `DASK_UCX__NET_DEVICES` entirely starting from UCX 1.10, because UCX is now capable of ensuring it uses the IB interface topologically closest to the GPU (which is what `"auto"` does now).

charlesbluca replied:
In this case, should I remove the recommendation to explicitly set `ucx.net-devices` to an ethernet device when InfiniBand is disabled? Seems like an inconsistency, unless I'm missing something.

pentschev replied:
The way I read is we recommend users to set to the "optimal" device, so I would still leave it in place. I know in the past we had issues with suboptimal devices being chosen and it seems like UCX 1.10+ is better in that regard, so what I meant to say was that starting from UCX 1.10 we'll then prefer not to set `UCX_NET_DEVICES`, but I don't think in our current state (UCX 1.9), we're ready for that.

pentschev commented on docs/source/ucx.rst:
```diff
@@ -1,35 +1,87 @@
+
+.. note::
+    These options can also be used with mainline Dask/Distributed.
+    However, this may disable a variety of features, such as *put some features exclusive to Dask-CUDA here*. 
```
`net-devices="auto"` is an exclusive feature of Dask-CUDA, I don't think there are others that apply to this section. The line below this should also be clarifying in this context.

charlesbluca replied:
Yeah, so far from the config section alone I have:

- Use of CUDA transfers for task on GPU
- Setting `net-devices="auto"`

Anything I'm missing here?

pentschev replied:
> Use of CUDA transfers for task on GPU

This is again not true. A user can very much use mainline Dask with GPUs, they will only not benefit from the specializations Dask-CUDA provides -- thus my comment on the line that is immediately below this clarifying that difference. Therefore, the only thing I can think of here is indeed `net-devices="auto"`.

pentschev reviewed (commented):
I added some comments here @charlesbluca , but it's looking good so far, thanks for working on this!

charlesbluca commented on docs/source/ucx.rst:
```diff
@@ -1,35 +1,87 @@
+Communication can be a major bottleneck in distributed systems.
+Dask-CUDA addresses this by supporting integration with `UCX <https://www.openucx.org/>`_, an optimized communication framework that provides high-performance networking and supports a variety of transport methods, including:
 
+- NVLink for *something ubiquitous to NVLink systems*
+- InfiniBand for *something ubiquitous to InfiniBand systems*
```
Sure! More details would probably be better suited for a hardware requirements section.

charlesbluca commented on docs/source/ucx.rst:
```diff
@@ -1,35 +1,87 @@
+Software
+^^^^^^^^
+
+*Does UCX integration require anything else other than UCX-Py?*
```
Thanks for the link! I will also go into some detail about when you would want to use the Conda package vs building/installing from source (essentially whether or not you plan to use IB/RDMA).

pentschev replied:
I think that information is better suited to the UCX-Py docs link I mentioned above. Let's try to avoid duplicating information which will eventually become inconsistent between the two docs.

charlesbluca commented on docs/source/ucx.rst:
```diff
@@ -1,35 +1,87 @@
+- ``ucx.tcp: true`` -- **required.**
+
+    Adds ``tcp`` to ``UCX_TLS``, enabling TCP transfers over UCX if NVLink or InfiniBand are unavailable or disabled.
+    *Why is this required if they are available and enabled?*
```
Would it maybe be better to use that for the reasoning to enable it in this case?

pentschev replied:
You can write a brief comment, but we probably don't want to rant about it here for too long as it's hard to draw a line of how much detail we want to go in. Here again a long rant would probably suit better UCX-Py (not suggesting we need to do it).

charlesbluca commented on docs/source/ucx.rst:
```diff
@@ -1,35 +1,87 @@
+- ``ucx.rdmacm: true`` -- recommended for InfiniBand.
+
+    Replaces ``sockcm`` with ``rdmacm`` in ``UCX_TLS`` and ``UCX_SOCKADDR_TLS_PRIORITY``, *enabling remote direct memory access (RDMA) for connection management.*
+    This is recommended by UCX for use with InfiniBand, and will have no effect if InfiniBand tranfers are disabled.
```
Is that to say that unexpected behavior will occur if `ucx.rdmacm` is enabled but `ucx.infiniband` is not?

pentschev replied:
Not unexpected behavior, the behavior is very much expected, resulting in an error on initialization.

charlesbluca commented:
Thanks for the review @pentschev! Addressed them as best I could.

I reorganized the use cases into a general "Usage" section and consolidated most of the information on client setup to an individual "Client" section that showcases `dask_cuda.initialize`; feel free to let me know how that looks. My first thought is that it may be nested too deep when it gets to sections on setting up the scheduler + workers using the CLI.

charlesbluca commented:
Is there anything we can do with the API reference to make `dask_cuda.initialize` show up with its arguments? It seems like it's meant for general use, but right now it seems like the only way to see what arguments it takes is to check the source code on GitHub.

charlesbluca commented:
Made some changes to the API source so that `dask_cuda.initialize.initialize` should show up properly now; also moved the docstring for the `dask_cuda.initialize` module to the function itself and added some parameters.

pentschev commented on dask_cuda/initialize.py:
```diff
@@ -44,6 +19,65 @@ def initialize(
+              - dask_cuda.initialize
+
+    See https://docs.dask.org/en/latest/configuration.html for more information
+    about Dask configuration.
```
We need to add a note here now to let users know that this is already done for `LocalCUDACluster` (including the client, if it's instantiated on the same process as the cluster) and `dask-cuda-worker`, so users don't need to do that again. The common place where users would want to use this is when they instantiate their standalone client connecting to a cluster that has already been started with a combination of `dask-scheduler`/`dask-cuda-worker`s.

charlesbluca replied:
Okay, in that case I will emphasize in the docstring that this script is meant for use with the client to mirror the pre-configured CUDA cluster/workers, and that the worker preload script would only need to be used with mainline Dask/Distributed.

pentschev commented on docs/source/ucx.rst:
```diff
@@ -1,135 +1,199 @@
 
-Note above how we use ``DASK_UCX__NET_DEVICES=mlx5_0:1`` (the Mellanox name for ``ib0``) and the same interface with ``--interface ib0``. If the system doesn't have an InfiniBand interface available, you would normally use the main network interface, such as ``eth0``, as seen below:
+  Replaces ``sockcm`` with ``rdmacm`` in ``UCX_TLS`` and ``UCX_SOCKADDR_TLS_PRIORITY``, enabling remote direct memory access (RDMA) for InfiniBand transfers.
+  This is recommended by UCX for use with InfiniBand, and will not work it is disabled.
```
```suggestion
  This is recommended by UCX for use with InfiniBand, and will not work if InfiniBand is disabled.
```

pentschev commented on docs/source/ucx.rst:
```diff
@@ -1,135 +1,199 @@
+
+.. note::
+    These options can be used with mainline Dask/Distributed.
+    However, this will disable a variety of features, such as the automatic detection of InfiniBand interfaces. 
```
```suggestion
    However, some features are exclusive to Dask-CUDA, such as the automatic detection of InfiniBand interfaces. 
```

pentschev commented on docs/source/ucx.rst:
```diff
@@ -1,135 +1,199 @@
-All ``DASK_*`` configurations described above have analogous parameters in ``dask-cuda-worker`` which are preferred over the regular configurations used for ``dask-scheduler`` due to some specializations, such as ``--net-devices="auto"`` which will correctly assign the topologically closest IB interface to the GPU of each worker, something that's not possible with ``DASK_UCX__NET_DEVICES``.
+A Dask cluster with UCX support can be started using the ``dask-cuda-worker`` CLI tool with a Dask scheduler which has been configured for UCX.
+
+*Do we want to include any reasons you would do this instead of using LocalCUDACluster?*
```
`LocalCUDACluster` is single-node only, anything else requires `dask-cuda-worker`.

pentschev commented on docs/source/ucx.rst:
```diff
@@ -1,135 +1,199 @@
+    DASK_RMM__POOL_SIZE=1GB \
+    dask-scheduler --protocol ucx --interface ib0
 
+Note the specification of ``mlx5_0:1`` as our UCX net device; because the scheduler does not rely upon Dask-CUDA, it cannot automatically detect InfiniBand interfaces, so we must specify one explicitly *(is this true?)*.
```
Yes, it is true.

pentschev commented on docs/source/ucx.rst:
```diff
@@ -1,135 +1,199 @@
+All the relevant Dask configuration options for UCX have analogous parameters in ``dask-cuda-worker``; see :doc:`Worker <worker>` for a complete list of these options.
 
-    from distributed import Client
+To start workers with all supported transports and a 1 gigabyte RMM pool *(do we need to specify an interface here?)*:
```
In most cases, no, UCX will be capable of identifying the optimal network interface to use.

pentschev commented on docs/source/ucx.rst:
```diff
@@ -1,135 +1,199 @@
     client = Client(cluster)
 
-    # Client code goes here
+Note the specification of ``"mlx5_0:1"`` as our net device in ``initialize()``; because the scheduler and client do not rely upon Dask-CUDA, they cannot automatically detect InfiniBand interfaces, so we must specify one explicitly (is this true?).
```
Yes, as in the previous comment in the scheduler section.

pentschev reviewed (commented):
Thanks @charlesbluca , this is looking good so far!

I added a few minor suggestions and replied most of your questions (if I missed any, please let me know).

Since this is getting larger than I initially predicted, what do you say we merge this after my comments are addressed and continue with samples in a follow-up PR? I generally think this is better to avoid getting PRs too large and increasingly more difficult to track changes.

charlesbluca commented:
Addressed your comments - let me know if I should change anything. If not, I'm happy to merge and handle the examples in a separate PR! Thanks for the help 😄

pentschev commented on dask_cuda/initialize.py:
```diff
@@ -25,10 +25,13 @@ def initialize(
     To ensure UCX works correctly, it is important to ensure it is initialized with
-    the correct options. This is important for scheduler, workers, and client. This
-    function will ensure that based on the flags and options passed by the user.
+    the correct options. This is especially important for the scheduler and client,
+    which cannot be configured to use UCX with arguments like ``LocalCUDACluster``
+    and ``dask-cuda-worker``. This function will ensure that they are provided a UCX
+    configuration based on the flags and options passed by the user.
```
But this function can't be used with `dask-scheduler`, so I don't think it really applies to that, only to the client. The scheduler will still have to rely on environment variables, so I would remove it from here.

charlesbluca replied:
If the scheduler still relies upon environment variables, does that mean that they must be set explicitly when running Python code?

I had initially removed the `os.environ[...]` calls from the `LocalCUDACluster` example as I thought all non-cluster/worker UCX config was handled through `initialize(...)` - I can add those back in if that's not the case.

pentschev replied:
We have 4 different types:

1. `LocalCUDACluster`: internally sets everything correctly to scheduler and workers, and also does that if `Client` is in the same process being instantiated after the cluster;
2. `dask-scheduler` (CLI): must set UCX options via environment variables (or `distributed.yaml`);
3. `dask-cuda-worker` (CLI): must  set UCX options via command-line arguments;
4. `Client` (standalone process): must set UCX options via environment variables (I think `distributed.yaml` also works here) or may use `dask_cuda.initialize`.

So what I meant was that for case 2 we can't use `dask_cuda.initialize` because it's a standalone process started via CLI. On the other hand, for `Client` we can either choose to use `dask_cuda.initialize` or environment variables, thus I think it only makes sense documenting for use with the client, so I would change the sentence above to:

"This is especially important for the ~scheduler and~ client, ..."

pentschev reviewed (changes requested):
One more small change, then I think we can merge it, but we're still blocked by https://github.com/rapidsai/dask-cuda/pull/546.

jakirkham commented:
rerun tests

pentschev reviewed (approved):
LGTM, thanks @charlesbluca for the work on improving docs here!

pentschev commented:
@gpucibot merge
